### ISSUE TITLE
Suggest to change default behaviour of `aggegate_region()` for components

### ISSUE THREAD

**Issue by danielhuppmann** [IAMconsortium/pyam#299]
## Introduction

The current implementation of [aggegate_region()](https://pyam-iamc.readthedocs.io/en/stable/api.html#pyam.IamDataFrame.aggregate_region) has an option to add `components` (other variables), which are only defined at the (aggregate) `region` level. Per default, the function uses all sub-categories of `variable` that are only defined at the `region` level (not in any of the `subregions`).

## Problem

The current implementation is very slow when iterating over many variables, and it might cause an issue when the sequence of aggregation matters on the outcome.

## Suggested solution

Allow `variable` to be a list of variables rather than only one (string), so that the region-aggregation can be performed in bulk. But this will cause problems with the automated detection of `components`. Also, passing the components for individual variables in the list would be cumbersome... (probably require a dictionary, what if that is not complete, ...).

Therefore, I suggest the following API change:
 - `variable` is a string or a list of strings, `components` is False (default): do not add any components, perform region aggregation on single variable or entire list (per variable)
 - `variable` is a string, `components` is True: automatically detect sub-categories (current behaviour)
 - `variable` is a string, `components` is a list: use `components` for the region-only-level values to be added
 - if `variable` is a list, `components` must be False

## Required changes in existing scripts

To continue current behaviour, existing scripts would need to add `components=True`.

**Comment by danielhuppmann** [comment]
@znicholls, you added this feature, in particular the automated detection of components. What are your thoughts here?

**Comment by znicholls** [comment]
I think the switch to having `components` be `False` by default is wise (especially because you can do that without affecting the `check_internal_consistency` API). 

One question, if you use `aggregate_region` with a list of variables, what are you actually doing? Shouldn't you be using `aggregate` and then `aggregate_region`?

**Comment by danielhuppmann** [comment]
We might want to have a similar discussion for [aggregate()](https://pyam-iamc.readthedocs.io/en/stable/api.html#pyam.IamDataFrame.aggregate) at some point, if this function should also take a list of variables (instead the current API taking only a single variable). This might be a more elaborate discussion, though...

However, for the AR6 upload-and-processing feature, we assume that the variable tree is complete, so no need to aggregate any variables from its disaggregated components; i.e., any scenario must explicitly provide `Emissions|CO2`, `Emissions|CO2|FFI`, `Emissions|CO2|AFOLU`, etc. (There are some arguments why this is a good idea).

In contrast, the regional-aggregation from native model regions (e.g., 11 in the case of MESSAGEix-GLOBIOM) to the IPCC-standard R5 regions is done as part of the AR6-Scenario-Explorer upload-and-postprocessing workflow, so we need a more efficient implementation of `aggregation_region()` to increase the processing speed.

**Comment by znicholls** [comment]
tl;dr: API change looks good to me as long as the behaviour of `check_internal_consistency` is unchanged.

------

Is the key goal here to speed things up?

> We might want to have a similar discussion for [aggregate()](https://pyam-iamc.readthedocs.io/en/stable/api.html#pyam.IamDataFrame.aggregate) at some point, if this function should also take a list of variables (instead the current API taking only a single variable). This might be a more elaborate discussion, though...

It already does take a list of variables no? They just go in `components`? If you had multiple variables going into `aggregate()` then how would you know how to map from `components` to `variables` (is that where your comments about a `dict` come in)? 

> In contrast, the regional-aggregation from native model regions (e.g., 11 in the case of MESSAGEix-GLOBIOM) to the IPCC-standard R5 regions is done as part of the AR6-Scenario-Explorer upload-and-postprocessing workflow, so we need a more efficient implementation of `aggregation_region()` to increase the processing speed.

If you're just going for speed, and you can be sure that you've sorted Bunkers out properly (so that the automatic detection of components doesn't matter), then I think the API change looks good.

**Comment by danielhuppmann** [comment]
> It already does take a list of variables no? They just go in components? If you had multiple variables going into aggregate() then how would you know how to map from components to variables (is that where your comments about a dict come in)?

Correct, it only takes one variable and a list of components - and having multiple variable with distinct components would require something more elaborate.

Therefore, my suggestion:
 - per default, ignore components irrespective of whether `variable` is a single item or a list of items (so that the default does not depend on how the function is called)
 - if components is important, the user has to call the function separately for each variable 
 - we can think about more improvements later

**Comment by znicholls** [comment]
sounds good to me
